Feature/testing

SofiaSeniv opened the pull request.

SofiaSeniv committed:
Add tests for UserService

SofiaSeniv committed:
fix: resolve authorization issue for roles

The issue was caused by the absence of roles in the token's claims, which led to incorrect authorization behavior.

SofiaSeniv committed:
feat: add tests for validation LoginDto and RegisterDro

SofiaSeniv committed:
fix: update test method to reflect changes in token generation

SofiaSeniv committed:
style: rename tests to follow naming conventions

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+        private readonly Mock<ITokenService> _mockTokenService;
+        private readonly Mock<ICustomUserManager> _mockCustomUserManager;
+        private readonly Mock<ICustomRoleManager> _mockCustomRoleManager;
+        private readonly Mock<IMapper> _mockMapper;
```
We don't mock mappers bc mapping is not business logic but only a projection

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+
+            _userService = new UserService(_mockTokenService.Object, _mockCustomUserManager.Object, _mockCustomRoleManager.Object, _mockMapper.Object);
+
+            _loginDtoValidator = new LoginDtoValidator();
+            _registerDtoValidator = new RegisterDtoValidator();
```
Should not belong to this test class - create 2 other test classes for each validator and move tests there

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+        public async Task LoginAsync_ShouldReturnJwtToken_WhenEmailAndPasswordAreValid()
+        {
+            //Arrange
+            var email = "test@example.com";
+            var password = "Password123*";
+            var loginDtoUser = new LoginDto { Email = email, Password = password };
+            var user = new User { Email = email, PasswordHash = password };
+            var roles = new List<string>{ Roles.Admin };
```
A lot of code duplication, apply DRY principle

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+        }
+
+        [Fact]
+        public async Task LoginAsync_ShouldReturnJwtToken_WhenEmailAndPasswordAreValid()
```
Remove AAA comments bc it's obvious for programmers which section of the test relates to which section of AAA

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+            var result = await _userService.LoginAsync(loginDtoUser);
+
+            //Assert
+            result.Should().Be("validJwtToken");
+            _mockCustomUserManager.Verify(userManager => userManager.FindByEmailAsync(loginDtoUser.Email), Times.Once());
+            _mockCustomUserManager.Verify(userManager => userManager.CheckPasswordAsync(user, loginDtoUser.Password), Times.Once);
+            _mockCustomUserManager.Verify(userManager => userManager.GetRolesAsync(user), Times.Once);
+            _mockTokenService.Verify(tokenService => tokenService.GenerateToken(user, roles), Times.Once);
```
You did too much assertions. Your test says ShouldReturnJwtToken, so only 1 assertion is enough here. For other Verify u can create tests like `ShouldFindUserByEmail`, `ShouldCheckPassword`, etc

sawabond replied:
BC if any of these extra assertions for your test fail, you will see failing `LoginAsync_ShouldReturnJwtToken_WhenEmailAndPasswordAreValid`, and u will think that the logic doesn't return JWT token, even though it is returning JWT

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+        public async Task LoginAsync_ShouldReturnNull_WhenPasswordIsInvalid()
+        {
+            //Arrange
+            var email = "test@example.com";
+            var password = "WrongPassword123*";
+            var loginDtoUser = new LoginDto { Email = email, Password = password };
+            var user = new User { Email = email, PasswordHash = password };
+            var roles = new List<string> { Roles.Admin };
+
+            _mockCustomUserManager.Setup(userManager => userManager.FindByEmailAsync(loginDtoUser.Email)).ReturnsAsync(user);
+            _mockCustomUserManager.Setup(userManager => userManager.CheckPasswordAsync(user, loginDtoUser.Password)).ReturnsAsync(false);
+
+            //Act
+            var result = await _userService.LoginAsync(loginDtoUser);
+
+            //Assert
+            result.Should().BeNull();
+            _mockCustomUserManager.Verify(userManager => userManager.FindByEmailAsync(loginDtoUser.Email), Times.Once());
+            _mockCustomUserManager.Verify(userManager => userManager.CheckPasswordAsync(user, loginDtoUser.Password), Times.Once);
+            _mockCustomUserManager.Verify(userManager => userManager.GetRolesAsync(user), Times.Never);
+            _mockTokenService.Verify(tokenService => tokenService.GenerateToken(user, roles), Times.Never);
```
Too much assertions and code duplication for setting up user

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+        public async Task LoginAsync_ShouldReturnNull_WhenUserNotFound()
+        {
+            //Arrange
+            var email = "test@example.com";
+            var password = "WrongPassword123*";
+            var loginDtoUser = new LoginDto { Email = email, Password = password };
+            var user = new User { Email = email, PasswordHash = password };
+            var roles = new List<string> { Roles.Admin };
+
+            _mockCustomUserManager.Setup(userManager => userManager.FindByEmailAsync(loginDtoUser.Email)).ReturnsAsync((User)null);
+
+            //Act
+            var result = await _userService.LoginAsync(loginDtoUser);
+
+            //Assert
+            result.Should().BeNull();
+            _mockCustomUserManager.Verify(userManager => userManager.FindByEmailAsync(loginDtoUser.Email), Times.Once());
+            _mockCustomUserManager.Verify(userManager => userManager.CheckPasswordAsync(user, loginDtoUser.Password), Times.Never);
+            _mockCustomUserManager.Verify(userManager => userManager.GetRolesAsync(user), Times.Never);
+            _mockTokenService.Verify(tokenService => tokenService.GenerateToken(user, roles), Times.Never);
```
Too much assertions and code duplication for setting up user

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+            var validationResult = await _loginDtoValidator.ValidateAsync(loginDtoUser);
+            var roles = new List<string> { Roles.Admin };
+
+            if (!validationResult.IsValid)
+            {
+                //Assert
+                _mockCustomUserManager.Verify(userManager => userManager.FindByEmailAsync(loginDtoUser.Email), Times.Never());
+                _mockCustomUserManager.Verify(userManager => userManager.CheckPasswordAsync(It.IsAny<User>(), loginDtoUser.Password), Times.Never);
+                _mockCustomUserManager.Verify(userManager => userManager.GetRolesAsync(It.IsAny<User>()), Times.Never);
+                _mockTokenService.Verify(tokenService => tokenService.GenerateToken(It.IsAny<User>(), roles), Times.Never);
+
+                return;
+            }
```
Test should not have conditions ever

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+            _mockTokenService.Verify(tokenService => tokenService.GenerateToken(user, roles), Times.Never);
+        }
+
+        [Theory]
+        [InlineData("valid.email@example.com", "Valid1@password")] // Happy path
+        [InlineData("", "")] // Both fields empty
+        [InlineData("invalidemail", "Valid1@password")] // Invalid email format
+        [InlineData("valid.email@example.com", "")] // Empty password
+        [InlineData("", "Valid1@password")] // Empty email
+        [InlineData("valid.email@example.com", "short")] // Password too short
+        [InlineData("valid.email@example.com", "aVeryVeryLongPasswordThatExceedsTheMaximumLengthAllowed1234567890@")] // Password too long
+        [InlineData("valid.email@example.com", "nouppercase1@")] // No uppercase
+        [InlineData("valid.email@example.com", "NOLOWERCASE1@")] // No lowercase
+        [InlineData("valid.email@example.com", "NoNumber@")] // No number
+        [InlineData("valid.email@example.com", "NoSpecialCharacter1")] // No special character
+        [InlineData("valid.email@example.com", "No Space1@")] // Contains spaces
+        [InlineData("email@domain.com", "Pass1!word")] // Another valid happy path
+        public async Task LoginAsync_ShouldNotProceed_WhenValidationFails(string email, string password)
```
Move this test to other test class

sawabond commented on BookingService/BookingService.Testing/UserServiceTests.cs:
```diff
@@ -0,0 +1,315 @@
+            //Arrange
+            var registerUserDto = new RegisterDto { Username = username, Email = email, Password = password, Role = role };
+            var user = new User();
+            var validationResult = _registerDtoValidator.Validate(registerUserDto);
+
+            if (!validationResult.IsValid)
+            {
+                _mockCustomUserManager.Verify(userManager => userManager.CreateAsync(user, registerUserDto.Password), Times.Never);
+                _mockCustomRoleManager.Verify(roleManager => roleManager.RoleExistsAsync(registerUserDto.Role), Times.Never);
+                _mockCustomUserManager.Verify(userManager => userManager.AddToRoleAsync(user, registerUserDto.Role), Times.Never);
+
+                return;
+            }
```
It obviously will be called never, because u apply these assertions BEFORE calling `RegisterAsync`.
The same: remove if condition and move validation assertions to other classes

SofiaSeniv committed:
refactor: migrate validation tests to a separate test classes

Migrating from UserServiceTests to LoginDtoValidatorTests and RegisterDtoValidatorTests

SofiaSeniv committed:
refactor: refactoring UserServiceTests class

remove unnecessary comments
consolidate duplicate code
refactor long functions into smaller ones
